[Issue #1] Application crashes (immediately) after choosing 'View log' from the menu.

amaa-99 opened the pull request:
Changed the required version to org.ocpsoft.prettytime:prettytime to 4.0.6.Final (as per the library documentation). This should be compatible with all the target OS versions.

ukanth reviewed (commented):
Rather than reverting to older version, can you please explain from which class the issue occurs ? we can try fixing with latest version of prettytime.

amaa-99 commented:
Thanks for having a look.

Please refer to to original issue on my fork:
https://github.com/amaa-99/afwall/issues/1

"This is due to the included version of the prettytime library using java 1.8 features which are not available in these OS versions.
As per the library documentation (https://github.com/ocpsoft/prettytime) version 4.0.6.Final or lower is to be used in these cases."

ukanth commented:
Java 1.8 (Java 8( is not Java 18. It should technically work with latest/old Android versions. I would like to fix the actual issue with 5.x version rather than downgrade it to 4.x

Commit:
* [Issue #1] Application crashes (immediately) after choosing ' View log' from the menu.
Dropped the use of the org.ocpsoft.prettytime:prettytime library: This fixes the crashes due to the library using unsupported language features on older platforms.

Commit:
* [Issue #1] Application crashes (immediately) after choosing ' View log' from the menu.
Addressed a couple of lint warnings.

amaa-99 commented:
I've removed the org.ocpsoft.prettytime:prettytime (which was used in a single spot to format a string), and replaced the implementation to use standard APIs.
(I've removed the related unused code and declarations which made it seem that that was used more often.)

amaa-99 commented on app/src/main/java/dev/ukanth/ufirewall/log/LogRecyclerViewAdapter.java:
```diff
@@ -107,42 +98,37 @@ private Bitmap getAppIcon26(PackageManager mPackageManager, ApplicationInfo appl
+            Calendar calendar = Calendar.getInstance();
+            calendar.setTimeInMillis(data.getTimestamp());
+            DateFormat dateFormat = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT);
+            holder.lastDenied.setText(dateFormat.format(calendar.getTime()));
```
This is the only place where that is used.

ukanth replied:
Thanks for PR. Does it retain the same functionality as prettytime ?

amaa-99 replied:
That's one of the java standard ways of printing a localized date/time (taking the time zone and ... into consideration).
It should do what's needed for the date/time string shown in the log view.

ukanth replied:
Actually it was not showing just data and time .It was showing the time lapse like 2s ago, 10s ago, 1 min ago etc.,

amaa-99 replied:
Note: In the detail view of the log entries the date/time is shown without any localization.
Probably the localized time should be shown there too.

amaa-99 replied:
I think the focus should be kept on fixing the crash first without adding complexity.
The pretty format can always be added on top...
I don't have much passion on which direction the solution goes, but I find it a pity having the app crash for something like that.
In my opinion as long as one can see the date/time of the logged event it's perfect (and it's even localized, so the US won't be mixing months with days ;)). What do you think?

ukanth replied:
Agree with your point. But I don't see any crashing reports based on this.

ukanth commented on app/src/main/java/dev/ukanth/ufirewall/log/LogRecyclerViewAdapter.java:
```diff
@@ -107,42 +98,37 @@ private Bitmap getAppIcon26(PackageManager mPackageManager, ApplicationInfo appl
-            info = Api.getPackageDetails(context, listMaps, data.getUid());
-            if(info!=null && info.applicationInfo !=null) {
-                Object drawable = info.applicationInfo.loadIcon(manager);
-                if (drawable instanceof Bitmap)
```
Why this logic is removed ?

amaa-99 replied:
The ' drawable instanceof Bitmap' is always false (note that loadIcon() always returns a Drawable).

ukanth replied:
It was added due to a crash on a specific devices.

amaa-99 replied:
Hummm, it probably fixed the crash by always falling into the else case.
I've noticed some inconsistencies around that area of the code related to different handling on different API levels.

amaa-99 replied:
I'm going to revert that change on this branch and make a new pull for it.

amaa-99 commented on app/src/main/java/dev/ukanth/ufirewall/log/LogRecyclerViewAdapter.java:
```diff
@@ -107,42 +98,37 @@ private Bitmap getAppIcon26(PackageManager mPackageManager, ApplicationInfo appl
+                holder.icon.setBackground(drawable);
+            } else {
                 Drawable appIcon = context.getDrawable(R.drawable.ic_unknown);
                 holder.icon.setImageBitmap(Api.getBitmapFromDrawable(appIcon));
```
The
```
holder.icon.setBackground(drawable);
```
is not consistent with the similar code a few lines below:
```
if (Build.VERSION.SDK_INT > android.os.Build.VERSION_CODES.O) {
    holder.icon.setBackground(context.getDrawable(R.drawable.ic_unknown));
} else {
    Drawable appIcon = context.getDrawable(R.drawable.ic_unknown);
    holder.icon.setImageBitmap(Api.getBitmapFromDrawable(appIcon));
}
```
At first glance it would seem that
```
holder.icon.setImageBitmap(Api.getBitmapFromDrawable(drawable))
```
should be used at that spot instead.

ukanth commented:
I will accept this PR if you handle the exception thrown by prettytime and fall back to normal timestamp rather than completely removing the feature.

amaa-99 commented:
Could you follow-up on this one? Just cherry pick what's useful into your branch and add the rest...
The issue can be reproduced in the device emulators for the Android 5 and 6: You just need to open the log view with some entries in there.
(Alternatively just add a call to those PrettyTime methods somewhere, e.g. in the creation of the apps list, to see the exception).

amaa-99 commented:
p.s. This seems to be a duplicate of the [(https://github.com/ukanth/afwall/issues/1344)] issue.

ukanth commented:
> Could you follow-up on this one? Just cherry pick what's useful into your branch and add the rest... The issue can be reproduced in the device emulators for the Android 5 and 6: You just need to open the log view with some entries in there. (Alternatively just add a call to those PrettyTime methods somewhere, e.g. in the creation of the apps list, to see the exception).

Thanks. As indicated above, please add an exception to handle rather than removing prettytime all together. I will merge the PR once that's done.

Commit:
* [Issue #18] The date/time formatting is not consistent across all views.
Corrected the date/time format in the details view (e.g. 19:00 would get displayed as '7:00').
Changed the detail view to not show the host name field in case the host name is not known.

Commit:
* [Issue #18] The date/time formatting is not consistent across all views.
Removed commented out code and unused methods.

ukanth commented:
Kindly resolve the conflicts and make a new PR. Thanks !
